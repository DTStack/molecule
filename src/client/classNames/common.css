*:focus-visible {
    outline: none;
}

body [tabindex='0']:focus,
body [tabindex='-1']:focus,
body .synthetic-focus,
body select:focus,
body input[type='button']:focus,
body input[type='text']:focus,
body button:focus,
body textarea:focus,
body input[type='search']:focus,
body input[type='checkbox']:focus {
    outline-width: 1px;
    outline-style: solid;
    outline-offset: -1px;
    outline-color: var(--focusBorder);
    opacity: 1;
}

body .synthetic-focus :focus {
    outline: 0 !important; /* elements within widgets that draw synthetic-focus should never show focus */
}

body .monaco-inputbox.info.synthetic-focus,
body .monaco-inputbox.warning.synthetic-focus,
body .monaco-inputbox.error.synthetic-focus,
body .monaco-inputbox.info input[type='text']:focus,
body .monaco-inputbox.warning input[type='text']:focus,
body .monaco-inputbox.error input[type='text']:focus {
    outline: 0 !important; /* outline is not going well with decoration */
}

body .monaco-list:focus {
    outline: 0 !important; /* tree indicates focus not via outline but through the focused item */
}

body {
    font-family: -apple-system, BlinkMacSystemFont, 'Segoe UI', 'Noto Sans', Helvetica, Arial,
        sans-serif, 'Apple Color Emoji', 'Segoe UI Emoji';
}
