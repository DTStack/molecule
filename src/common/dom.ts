export type HTMLElementType = HTMLElement | null;
export type TriggerEvent = 'click' | 'contextmenu' | 'hover';
/**
 * specify `rightBottom` means align to the bottom and keep in right
 */
export type PlacementType = 'top' | 'right' | 'bottom' | 'left' | 'rightBottom';
export const select = document.querySelector.bind(document);
export const selectAll = document.querySelectorAll.bind(document);

export interface IPosition {
    x: number;
    y: number;
}

/**
 * Get Document Rectangle info
 */
export function getDocumentRect() {
    const body = document.body;
    const html = document.documentElement;

    const clientWidth = html.clientWidth;
    const clientHeight = html.clientHeight;

    const height = Math.max(
        body.scrollHeight,
        body.offsetHeight,
        clientHeight,
        html.scrollHeight,
        html.offsetHeight
    );

    const width = Math.max(
        body.scrollWidth,
        body.offsetWidth,
        clientWidth,
        html.scrollWidth,
        html.offsetWidth
    );

    return {
        height,
        width,
        clientWidth,
        clientHeight,
    };
}

/**
 * Returns the position of element relative to element position
 * @param element target element
 * @param relativePos the relative element position
 */
export function getRelativePosition(
    element: HTMLElement,
    relativePos: IPosition
) {
    const { clientWidth, clientHeight } = getDocumentRect();
    const anchorX = relativePos.x;
    const anchorY = relativePos.y;

    const marginRight = clientWidth - anchorX;
    const marginBottom = clientHeight - anchorY;

    const viewHeight = element.offsetHeight;
    const viewWidth = element.offsetWidth;

    const x = marginRight < viewWidth ? anchorX - viewWidth : anchorX;
    const y = marginBottom < viewHeight ? anchorY - viewHeight : anchorY;

    return {
        x,
        y,
    };
}

export function getEventPosition(e: React.MouseEvent) {
    return {
        x: e.clientX,
        y: e.clientY,
    };
}

export function findParentByClassName<T>(element, className): T | null {
    try {
        while (element && element) {
            const classes = element.getAttribute('class');
            if (classes && classes.indexOf(className) > -1) {
                return element;
            }
            element = element.parentElement;
        }
    } catch (e) {
        throw new Error(`findParentByClassName error: ${e}`);
    }
    return null;
}

export function triggerEvent(trigger: TriggerEvent) {
    switch (trigger) {
        case 'click': {
            return 'onClick';
        }
        case 'hover': {
            return 'onMouseOver';
        }
        case 'contextmenu': {
            return 'onContextMenu';
        }
        default: {
            return trigger;
        }
    }
}

/**
 * Get the element position by placement and DOMRect
 * @param placement top | right | bottom | left
 * @param domRect Dom rect info, normally get it from getBoundingClientRect function
 */
export function getPositionByPlacement(
    placement: PlacementType,
    domRect: DOMRect
): IPosition {
    let x = domRect.x; // Initial placement is top
    let y = domRect.y;

    if (placement === 'top') {
        y = domRect.y - domRect.height;
    } else if (placement === 'right') {
        x = domRect.x + domRect.width;
    } else if (placement === 'bottom') {
        y = domRect.y + domRect.height;
    } else if (placement === 'left') {
        x = domRect.x - domRect.width;
    } else if (placement === 'rightBottom') {
        x = domRect.x + domRect.width;
        y = domRect.y + domRect.height;
    }
    return { x, y };
}

export function getAttr(domElement: HTMLElement, attr) {
    return domElement.getAttribute(attr) || '';
}

/**
 * Get an element the center coords
 * @param element HTMLElement
 * @returns
 */
export function getElementClientCenter(element: HTMLElement) {
    const { left, top, width, height } = element.getBoundingClientRect();
    return {
        x: left + width / 2,
        y: top + height / 2,
    };
}

/**
 * Get the data-* attributions from props
 * @param props
 * @returns
 */
export function getDataAttributionsFromProps(
    props: Record<string, any>
): Record<string, any> {
    return Object.keys(props).reduce((pre, cur) => {
        if (cur.startsWith('data-')) {
            pre[cur] = props[cur];
        }
        return pre;
    }, {});
}
