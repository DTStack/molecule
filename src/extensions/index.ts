import { ExtendsActions } from './actions';
import { ExtendsActivityBar } from './activityBar';
import { ExtendsAuxiliaryBar } from './auxiliaryBar';
import { ExtendsContextMenu } from './contextMenu';
import { ExtendsEditor } from './editor';
import { ExtendsEditorTree } from './editorTree';
import { ExtendsExplorer } from './explorer';
import { ExtendsFolderTree } from './folderTree';
import { LocalesExtends } from './locales';
import { ExtendsMenuBar } from './menuBar';
import { ExtendsNotification } from './notification';
import { ExtendsPanel } from './panel';
import { ExtendsSearch } from './search';
import { ExtendsSettings } from './settings';
import { ExtendsSidebar } from './sidebar';
import { ExtendsStatusBar } from './statusBar';
import { ColorThemeExtension } from './themes';

export default [
    LocalesExtends,
    ColorThemeExtension,
    ExtendsActions,
    ExtendsEditor,
    ExtendsActivityBar,
    ExtendsEditorTree,
    ExtendsPanel,
    ExtendsMenuBar,
    ExtendsNotification,
    ExtendsStatusBar,
    ExtendsSearch,
    ExtendsSettings,
    ExtendsFolderTree,
    ExtendsSidebar,
    ExtendsExplorer,
    ExtendsContextMenu,
    ExtendsAuxiliaryBar,
];
