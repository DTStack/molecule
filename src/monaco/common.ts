import { localize } from 'monaco-editor/esm/vs/nls';

export interface IDisposable {
    dispose(): void;
}

export enum KeybindingWeight {
    EditorCore = 0,
    EditorContrib = 100,
    WorkbenchContrib = 200,
    BuiltinExtension = 300,
    ExternalExtension = 400,
}

export const CATEGORIES = {
    View: { value: localize('view', 'View'), original: 'View' },
    Help: { value: localize('help', 'Help'), original: 'Help' },
    Preferences: {
        value: localize('preferences', 'Preferences'),
        original: 'Preferences',
    },
    Developer: {
        value: localize(
            {
                key: 'developer',
                comment: [
                    'A developer on Code itself or someone diagnosing issues in Code',
                ],
            },
            'Developer'
        ),
        original: 'Developer',
    },
};
