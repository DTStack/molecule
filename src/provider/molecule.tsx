import React, { useLayoutEffect } from 'react';
import create, { IConfigProps } from './create';

export default function Provider({
    defaultLocale,
    extensions,
    children,
}: IConfigProps & { children: React.ReactElement }) {
    useLayoutEffect(() => {
        const instance = create({
            defaultLocale,
            extensions,
        });

        instance.render(children);
    }, []);

    return children;
}
