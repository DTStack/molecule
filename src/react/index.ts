export * from './component';
export * from './helper';
export * from './connector';
export * from './controller';
