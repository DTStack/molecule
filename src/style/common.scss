$prefix: 'mo';

@function prefix($name, $cusPrefix: $prefix) {
    @return '.' + $cusPrefix + '-' + $name;
}

// ============= The Workbench parts =============
$actionBar: prefix('action-bar');
$btn: prefix('btn');
$checkbox: prefix('checkbox');
$collapse: prefix('collapse');
$rcCollapse: prefix('collapse', 'rc');
$name: prefix('context-view');
$dropDown: prefix('drop-down');
$input: prefix('input');
$list: prefix('list');
$menu: prefix('menu');
$subMenu: prefix('sub-menu');
$scrollbar: prefix('scrollbar');
$select: prefix('select');
$tabSwitcher: 'tab-switcher';
$toolBar: prefix('tool-bar');
$tree: prefix('tree');
$tabs: prefix('tabs');
$tab: prefix('tab');
$modal: prefix('modal');
$confirm: prefix('confirm');
$breadcrumb: prefix('breadcrumb');
$search: prefix('search');
$rcTooltip: prefix('tooltip', 'rc');
$split: prefix('split');
$bell: prefix('bell');

// ============= The Workbench core component =============
$compositeBar: prefix('compositeBar');
$editor: prefix('editor');
$menuBar: prefix('menuBar');
$panel: prefix('panel');
$settings: prefix('settings');
$sidebar: prefix('sidebar');
$statusBar: prefix('statusBar');
$workbench: prefix('workbench');
$mainBench: prefix('mainBench');
$activityBar: prefix('activityBar');
$notification: prefix('notification');
$problems: prefix('problems');
$output: prefix('output');
$folderTree: prefix('folderTree');
$editorTree: prefix('editorTree');

// The Naming of BEM Element
@function bem-ele($block, $element) {
    @return $block + '__' + $element;
}

// The Naming of BEM Modifier
@function bem-mod($blockOrElement, $modifier) {
    @return $blockOrElement + '--' + $modifier;
}
