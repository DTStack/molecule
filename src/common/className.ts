import { Utils } from '@dtinsight/dt-utils/lib';
import { isEmpty } from 'lodash';
import { APP_PREFIX } from 'mo/common/const';
/**
 * This function help you prefix a css class name, default is molecule.
 * Example: prefixClaName('test') will return 'molecule-test',
 * prefixClaName('test', 'c') will return 'c-test'
 * @param name Default class name
 * @param prefix The prefix of class name you want to append
 */
export function prefixClaName(name: string, prefix: string = APP_PREFIX) {
    return name ? `${prefix}-${name}` : '';
}

export function classNames(...args) {
    if (isEmpty(args)) return;
    const classList: string[] = [];
    for (const arg of args) {
        if (!arg) continue;
        const argType = typeof arg;
        if (argType === 'string' || argType === 'number') {
            classList.push(`${arg}`);
            continue;
        }
        if (argType === 'object') {
            if (arg.toString !== Object.prototype.toString) {
                classList.push(arg.toString());
                continue;
            }
            for (const key in arg) {
                if (Object.hasOwnProperty.call(arg, key) && arg[key]) {
                    classList.push(key);
                }
            }
        }
    }
    return classList.join(' ');
}
/**
 * Element names may consist of Latin letters, digits, dashes and underscores.
 * CSS class is formed as block name plus two underscores plus element name: .block__elem
 * @param block
 * @param element
 */
export function getBEMElement(block: string, element: string) {
    return `${block}__${element}`;
}

/**
 * CSS class is formed as block’s or element’s name plus two dashes:
 * .block--mod or .block__elem--mod and .block--color-black with .block--color-red.
 * Spaces in complicated modifiers are replaced by dash.
 * @param blockOrElement
 * @param modifier
 */
export function getBEMModifier(blockOrElement: string, modifier: string) {
    return `${blockOrElement}--${modifier}`;
}

/**
 * Returns the className of font-family in mac
 * @returns
 */
export function getFontInMac() {
    return Utils.isMacOs() ? 'mac' : '';
}
