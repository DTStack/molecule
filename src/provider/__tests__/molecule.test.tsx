import React from 'react';
import { render } from '@testing-library/react';
import '@testing-library/jest-dom';

import { select } from 'mo/common/dom';
import molecule, { MoleculeProvider, Workbench } from 'mo';

import { customExtensions } from '../../../stories/extensions';

describe('Test MoleculeProvider', () => {
    let original;
    beforeEach(() => {
        original = HTMLElement.prototype.getBoundingClientRect;
        // @ts-ignore
        HTMLElement.prototype.getBoundingClientRect = () => ({
            width: 500,
            height: 0,
        });

        // Reset the extensions loaded state
        molecule.extension.setLoaded(false);
    });

    afterEach(() => {
        HTMLElement.prototype.getBoundingClientRect = original;
    });

    test('Match The MoleculeProvider snapshot', () => {
        const { asFragment } = render(
            <MoleculeProvider>
                <Workbench />
            </MoleculeProvider>
        );
        expect(asFragment()).toMatchSnapshot();
    });

    test('MoleculeProvider should render built-in Workbench extensions', () => {
        render(
            <MoleculeProvider>
                <Workbench />
            </MoleculeProvider>
        );
        expect(
            select('div[data-id="sidebar.explore.title"]')
        ).toBeInTheDocument();
        expect(
            select('div[data-id="sidebar.search.title"]')
        ).toBeInTheDocument();
        expect(select('.mo-welcome')).toBeInTheDocument();
        expect(
            select('div[id="statusbar.problems.title"]')
        ).toBeInTheDocument();
    });

    test('MoleculeProvider load the extensions', async () => {
        render(
            <MoleculeProvider extensions={customExtensions}>
                <Workbench />
            </MoleculeProvider>
        );
        await expect(
            select('div[data-id="ActivityBarTestPane"]')
        ).toBeInTheDocument();
    });

    test('MoleculeProvider load the locale language extensions', () => {
        localStorage.removeItem('mo.localeId');

        render(
            <MoleculeProvider
                extensions={customExtensions}
                defaultLocale="zh-CN"
            >
                <Workbench />
            </MoleculeProvider>
        );
    });
});
