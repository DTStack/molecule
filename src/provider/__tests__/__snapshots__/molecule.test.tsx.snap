// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Test MoleculeProvider Match The MoleculeProvider snapshot 1`] = `
<DocumentFragment>
  <div
    class="mo"
    id="molecule"
    tabindex="0"
  >
    <div
      class="mo-workbench"
    >
      <div
        class="mo-mainBench"
      >
        <div
          class="mo-compositeBar"
        >
          <div
            class="mo-menuBar"
          >
            <div
              class="mo-drop-down mo-drop-down--right mo-menuBar__action"
            >
              <span
                class="codicon codicon-menu"
              />
            </div>
          </div>
          <div
            class="mo-activityBar"
            id="activityBar"
          >
            <div
              class="mo-activityBar__container"
            >
              <ul>
                <li
                  class="mo-activityBar__item mo-activityBar__item--checked"
                  id="sidebar.explore.title"
                >
                  <span
                    class="mo-activityBar__label codicon codicon-files"
                    title="Explorer"
                  />
                  <div
                    class="mo-activityBar__indicator"
                  />
                </li>
                <li
                  class="mo-activityBar__item"
                  id="sidebar.search.title"
                >
                  <span
                    class="mo-activityBar__label codicon codicon-search"
                    title="Search"
                  />
                </li>
              </ul>
              <ul
                class="mo-activityBar__global"
              >
                <li
                  class="mo-activityBar__item"
                  id="global.menu.account"
                >
                  <span
                    class="mo-activityBar__label codicon codicon-account"
                    title="Account"
                  />
                </li>
                <li
                  class="mo-activityBar__item"
                  id="global.menu.settings"
                >
                  <div
                    class="mo-drop-down mo-drop-down--rightBottom"
                  >
                    <span
                      class="mo-activityBar__label codicon codicon-settings-gear"
                      title="Settings"
                    />
                  </div>
                </li>
              </ul>
            </div>
          </div>
        </div>
        <div
          class="mo-split"
        >
          <div
            class="mo-split__sash"
          >
            <div
              class="mo-split__sash__item mo-split__sash__item--disabled mo-split__sash__item--vertical"
              role="Resizer"
              style="width: 4px; left: -2px;"
            />
            <div
              class="mo-split__sash__item mo-split__sash__item--vertical"
              role="Resizer"
              style="width: 4px; left: 298px;"
            />
          </div>
          <div
            class="mo-split__pane"
          >
            <div
              class="mo-split__pane__item mo-split__pane__item--visible"
              style="left: 0px; width: 300px;"
            >
              <div
                class="mo-sidebar"
              >
                <div
                  class="mo-sidebar__pane"
                  data-id="sidebar.explore.title"
                  style="visibility: visible; z-index: 1;"
                >
                  <div
                    class="mo-sidebar-explorer"
                  >
                    <header
                      class="mo-sidebar__header"
                    >
                      <div
                        class="mo-sidebar__title"
                      >
                        <h2>
                          Explorer
                        </h2>
                      </div>
                      <div
                        class="mo-sidebar__toolbar"
                      >
                        <div
                          class="mo-tool-bar"
                        >
                          <div
                            class="mo-action-bar"
                          >
                            <ul
                              class="mo-action-bar__container"
                            >
                              <li
                                class="mo-action-bar__item"
                                data-index="0"
                              >
                                <span
                                  class="codicon codicon-ellipsis"
                                />
                              </li>
                            </ul>
                          </div>
                        </div>
                      </div>
                    </header>
                    <div
                      class="mo-sidebar__content"
                    >
                      <div
                        class="mo-collapse"
                      >
                        <div
                          class="mo-split"
                        >
                          <div
                            class="mo-split__sash"
                          >
                            <div
                              class="mo-split__sash__item mo-split__sash__item--disabled mo-split__sash__item--horizontal"
                              role="Resizer"
                              style="height: 4px; top: -2px;"
                            />
                            <div
                              class="mo-split__sash__item mo-split__sash__item--disabled mo-split__sash__item--horizontal"
                              role="Resizer"
                              style="height: 4px; top: 24px;"
                            />
                            <div
                              class="mo-split__sash__item mo-split__sash__item--disabled mo-split__sash__item--horizontal"
                              role="Resizer"
                              style="height: 4px; top: 50px;"
                            />
                          </div>
                          <div
                            class="mo-split__pane"
                          >
                            <div
                              class="mo-split__pane__item mo-split__pane__item--visible mo-collapse__pane"
                              style="top: 0px; height: 26px;"
                            >
                              <div
                                class="mo-collapse__item"
                                data-collapse-id="sidebar.explore.openEditor"
                              >
                                <div
                                  class="mo-collapse__header"
                                  tabindex="0"
                                >
                                  <span
                                    class="codicon codicon-chevron-right"
                                  />
                                  <span
                                    class="mo-collapse__header__title"
                                  >
                                    Open Editors
                                  </span>
                                  <div
                                    class="mo-collapse__extra"
                                  />
                                </div>
                                <div
                                  class="mo-collapse__content"
                                  data-collapse-index="0"
                                  tabindex="0"
                                />
                              </div>
                            </div>
                            <div
                              class="mo-split__pane__item mo-split__pane__item--visible mo-collapse__pane"
                              style="top: 26px; height: 26px;"
                            >
                              <div
                                class="mo-collapse__item"
                                data-collapse-id="sidebar.explore.folders"
                              >
                                <div
                                  class="mo-collapse__header"
                                  tabindex="0"
                                >
                                  <span
                                    class="codicon codicon-chevron-right"
                                  />
                                  <span
                                    class="mo-collapse__header__title"
                                  >
                                    No Open Folder
                                  </span>
                                  <div
                                    class="mo-collapse__extra"
                                  />
                                </div>
                                <div
                                  class="mo-collapse__content"
                                  data-collapse-index="1"
                                  tabindex="0"
                                >
                                  <div
                                    data-content="sidebar.explore.folders"
                                  >
                                    <div
                                      style="padding: 10px 5px;"
                                    >
                                      you have not yet opened a folder
                                      <a
                                        class="mo-btn mo-btn--normal"
                                      >
                                        Add Folder
                                      </a>
                                    </div>
                                  </div>
                                </div>
                              </div>
                            </div>
                            <div
                              class="mo-split__pane__item mo-split__pane__item--visible mo-collapse__pane"
                              style="top: 52px; height: 26px;"
                            >
                              <div
                                class="mo-collapse__item"
                                data-collapse-id="outline"
                              >
                                <div
                                  class="mo-collapse__header"
                                  tabindex="0"
                                >
                                  <span
                                    class="codicon codicon-chevron-right"
                                  />
                                  <span
                                    class="mo-collapse__header__title"
                                  >
                                    Outline
                                  </span>
                                  <div
                                    class="mo-collapse__extra"
                                  />
                                </div>
                                <div
                                  class="mo-collapse__content"
                                  data-collapse-index="2"
                                  tabindex="0"
                                />
                              </div>
                            </div>
                          </div>
                        </div>
                      </div>
                    </div>
                  </div>
                </div>
                <div
                  class="mo-sidebar__pane"
                  data-id="sidebar.search.title"
                  style="visibility: hidden; z-index: -1;"
                >
                  <div
                    class="sidebar-search-pane"
                  >
                    <header
                      class="mo-sidebar__header"
                    >
                      <div
                        class="mo-sidebar__title"
                      >
                        <h2>
                          Search
                        </h2>
                      </div>
                      <div
                        className="mo-collapse__content"
                        data-content="sidebar.explore.openEditor"
                        tabIndex={0}
                      />
                    </div>
                    <div
                      className="mo-collapse__item"
                      data-content="sidebar.explore.folders"
                    >
                      <div
                        className="mo-collapse__header"
                        onClick={[Function]}
                        tabIndex={0}
                      >
                        <span
                          className="codicon codicon-chevron-right"
                        />
                        <span
                          className="mo-collapse__header__title"
                        >
                          Default Project Name
                        </span>
                        <div
                          className="mo-collapse__extra"
                        />
                      </div>
                      <div
                        className="mo-collapse__content"
                        data-content="sidebar.explore.folders"
                        tabIndex={0}
                      >
                        <div
                          class="mo-tool-bar"
                        >
                          <div
                            class="mo-action-bar"
                          >
                            <ul
                              class="mo-action-bar__container"
                            >
                              <li
                                class="mo-action-bar__item mo-action-bar__item--disabled"
                                data-index="0"
                              >
                                <span
                                  class="codicon codicon-refresh"
                                />
                              </li>
                              <li
                                class="mo-action-bar__item mo-action-bar__item--disabled"
                                data-index="1"
                              >
                                <span
                                  class="codicon codicon-clear-all"
                                />
                              </li>
                              <li
                                class="mo-action-bar__item mo-action-bar__item--disabled"
                                data-index="2"
                              >
                                <span
                                  class="codicon codicon-collapse-all"
                                />
                              </li>
                            </ul>
                          </div>
                        </div>
                      </div>
                    </header>
                    <div
                      class="mo-sidebar__content"
                    >
                      <div
                        class="mo-search"
                      >
                        <span
                          class="mo-search__replace__button codicon codicon-chevron-right"
                        />
                        <div
                          class="mo-search__group"
                        >
                          <div
                            class="mo-search__input mo-search__target"
                          >
                            <textarea
                              autocapitalize="off"
                              autocorrect="off"
                              class=""
                              placeholder="Search"
                              spellcheck="false"
                              title="Search"
                            />
                            <div
                              class="mo-action-bar mo-search__toolbar"
                            >
                              <ul
                                class="mo-action-bar__container"
                              >
                                <li
                                  class="mo-action-bar__item"
                                  data-index="0"
                                >
                                  <span
                                    class="codicon codicon-case-sensitive"
                                  />
                                </li>
                                <li
                                  class="mo-action-bar__item"
                                  data-index="1"
                                >
                                  <span
                                    class="codicon codicon-whole-word"
                                  />
                                </li>
                                <li
                                  class="mo-action-bar__item"
                                  data-index="2"
                                >
                                  <span
                                    class="codicon codicon-regex"
                                  />
                                </li>
                              </ul>
                            </div>
                          </div>
                        </div>
                      </div>
                      <div
                        class="mo-tree mo-search__tree"
                        draggable="false"
                        role="tree"
                      />
                    </div>
                  </div>
                </div>
              </div>
            </div>
            <div
              class="mo-split__pane__item mo-split__pane__item--visible"
              style="left: 300px; width: 200px;"
            >
              <div
                class="mo-split"
              >
                <div
                  class="mo-split__sash"
                >
                  <div
                    class="mo-split__sash__item mo-split__sash__item--disabled mo-split__sash__item--horizontal"
                    role="Resizer"
                    style="height: 4px; top: -2px;"
                  />
                  <div
                    class="mo-split__sash__item mo-split__sash__item--horizontal"
                    role="Resizer"
                    style="height: 4px; top: -2px;"
                  />
                </div>
                <div
                  class="mo-split__pane"
                >
                  <div
                    class="mo-split__pane__item"
                    style="top: 0px; height: 0px;"
                  >
                    <div
                      class="mo-editor"
                    >
                      <div
                        class="mo-welcome"
                      >
                        <span
                          style="font-size: 0px;"
                        >
                          <svg
                            class="logo"
                            height="1em"
                            viewBox="0 0 120 120"
                            width="1em"
                          >
                            <g
                              fill="none"
                              fill-rule="evenodd"
                              opacity="0.48"
                              stroke="none"
                              stroke-width="1"
                            >
                              <g
                                fill="#A2A2A2"
                                transform="translate(-660 -152)"
                              >
                                <g
                                  transform="translate(660 152)"
                                >
                                  <path
                                    d="M59.3 96.335c4.87 0 8.817 3.954 8.817 8.832S64.169 114 59.3 114c-4.87 0-8.818-3.955-8.818-8.833s3.948-8.832 8.818-8.832zm-22.444-6.826a3.21 3.21 0 013.206 3.212 3.21 3.21 0 01-3.206 3.212 3.21 3.21 0 01-3.206-3.212 3.21 3.21 0 013.206-3.212zm44.887 0a3.21 3.21 0 013.206 3.212 3.21 3.21 0 01-3.206 3.212 3.21 3.21 0 01-3.206-3.212 3.21 3.21 0 013.206-3.212zM59.3 72.245c4.87 0 8.817 3.955 8.817 8.833s-3.948 8.833-8.817 8.833c-4.87 0-8.818-3.955-8.818-8.833s3.948-8.833 8.818-8.833zm-42.483 0c4.87 0 8.817 3.955 8.817 8.833s-3.947 8.833-8.817 8.833S8 85.956 8 81.078s3.948-8.833 8.817-8.833zm85.366 0c4.87 0 8.817 3.955 8.817 8.833s-3.948 8.833-8.817 8.833c-4.87 0-8.817-3.955-8.817-8.833s3.947-8.833 8.817-8.833zm-65.327-6.022a3.21 3.21 0 013.206 3.212 3.21 3.21 0 01-3.206 3.212 3.21 3.21 0 01-3.206-3.212 3.21 3.21 0 013.206-3.212zm44.887 0a3.21 3.21 0 013.206 3.212 3.21 3.21 0 01-3.206 3.212 3.21 3.21 0 01-3.206-3.212 3.21 3.21 0 013.206-3.212zM59.3 48.156c4.87 0 8.817 3.955 8.817 8.833s-3.948 8.833-8.817 8.833c-4.87 0-8.818-3.955-8.818-8.833s3.948-8.833 8.818-8.833zm-42.483 0c4.87 0 8.817 3.955 8.817 8.833s-3.947 8.833-8.817 8.833S8 61.867 8 56.989s3.948-8.833 8.817-8.833zm85.366 0c4.87 0 8.817 3.955 8.817 8.833s-3.948 8.833-8.817 8.833c-4.87 0-8.817-3.955-8.817-8.833s3.947-8.833 8.817-8.833zM80.942 35.31c4.87 0 8.817 3.954 8.817 8.832s-3.948 8.833-8.817 8.833c-4.87 0-8.817-3.955-8.817-8.833s3.947-8.832 8.817-8.832zm-43.284 0c4.87 0 8.817 3.954 8.817 8.832s-3.948 8.833-8.817 8.833c-4.87 0-8.818-3.955-8.818-8.833s3.948-8.832 8.818-8.832zm64.525-11.242c4.87 0 8.817 3.954 8.817 8.833 0 4.878-3.948 8.832-8.817 8.832-4.87 0-8.817-3.954-8.817-8.832 0-4.879 3.947-8.833 8.817-8.833zm-85.366 0c4.87 0 8.817 3.954 8.817 8.833 0 4.878-3.947 8.832-8.817 8.832S8 37.778 8 32.9c0-4.879 3.948-8.833 8.817-8.833zm42.483 5.62a3.21 3.21 0 013.206 3.213 3.21 3.21 0 01-3.206 3.212 3.21 3.21 0 01-3.207-3.212 3.21 3.21 0 013.207-3.212zm-19.238-13.65a3.21 3.21 0 013.206 3.212 3.21 3.21 0 01-3.206 3.212 3.21 3.21 0 01-3.206-3.212 3.21 3.21 0 013.206-3.212zm38.475 0a3.21 3.21 0 013.206 3.212 3.21 3.21 0 01-3.206 3.212 3.21 3.21 0 01-3.206-3.212 3.21 3.21 0 013.206-3.212zM59.3 6a3.21 3.21 0 013.206 3.212 3.21 3.21 0 01-3.206 3.212 3.21 3.21 0 01-3.207-3.212A3.21 3.21 0 0159.3 6z"
                                  />
                                </g>
                              </g>
                            </g>
                          </svg>
                        </span>
                        <h1
                          class="title"
                        >
                          Molecule
                        </h1>
                        <div
                          class="keybindings"
                        >
                          <ul>
                            <li
                              class="keys"
                            >
                              <span>
                                Command Palette
                              </span>
                              <span />
                            </li>
                          </ul>
                        </div>
                      </div>
                    </div>
                  </div>
                  <div
                    class="mo-split__pane__item"
                    style="top: 0px; height: 0px;"
                  >
                    <div
                      class="mo-panel"
                    >
                      <div
                        class="mo-panel__header"
                      >
                        <div
                          class="mo-tabs mo-tabs--line"
                        >
                          <div
                            class="mo-tabs__header"
                          >
                            <div
                              class="mo-tab__item mo-tab__item--active"
                              draggable="true"
                            >
                              Problems
                            </div>
                            <div
                              class="mo-tab__item"
                              draggable="true"
                            >
                              output
                            </div>
                          </div>
                        </div>
                        <div
                          class="mo-action-bar mo-panel__toolbar"
                        >
                          <ul
                            class="mo-action-bar__container"
                          >
                            <li
                              class="mo-action-bar__item"
                              data-index="0"
                            >
                              <span
                                class="codicon codicon-chevron-up"
                              />
                            </li>
                            <li
                              class="mo-action-bar__item"
                              data-index="1"
                            >
                              <span
                                class="codicon codicon-close"
                              />
                            </li>
                          </ul>
                        </div>
                      </div>
                      <div
                        class="mo-panel__container"
                        tabindex="0"
                      >
                        <div
                          style="margin: 0px 18px; user-select: none;"
                        >
                          No problems have been detected in the workspace.
                        </div>
                      </div>
                    </div>
                  </div>
                </div>
              </div>
            </div>
          </div>
        </div>
      </div>
    </div>
  </div>
  <div
    style={
      Object {
        "display": undefined,
      }
    }
  >
    <div
      className="mo-statusBar"
      id="statusBar"
    >
      <div
        className="mo-statusBar__left-items"
      >
        <div
          className="mo-statusBar__item"
          id="statusbar.problems.title"
        >
          <a
            onClick={[Function]}
            tabIndex={-1}
            title="Problems"
          >
            <span
              className="codicon codicon-error"
            />
             0 
            <span
              className="codicon codicon-warning"
            />
             0 
            <span
              className="codicon codicon-info"
            />
             0
          </a>
        </div>
      </div>
      <div
        className="mo-statusBar__right-items"
      >
        <div
          className="mo-statusBar__item"
          id="MO_NOTIFICATION"
        >
          <a
            onClick={[Function]}
            tabIndex={-1}
            title="Notification"
          >
            <span
              className="codicon codicon-bell"
              onClick={[Function]}
            />
            <div
              className="mo-notification mo-context-view--shadow"
              style={
                Object {
                  "display": "none",
                }
              }
            >
              <header
                className="mo-notification__header"
              >
                <span>
                  No new notifications
                </span>
                <div
                  className="mo-action-bar"
                >
                  <ul
                    className="mo-action-bar__container"
                  >
                    <li
                      className="mo-action-bar__item"
                      data-index={0}
                      onClick={[Function]}
                    >
                      <span
                        className="codicon codicon-clear-all"
                        onMouseEnter={[Function]}
                        onMouseLeave={[Function]}
                      />
                    </li>
                    <li
                      className="mo-action-bar__item"
                      data-index={1}
                      onClick={[Function]}
                    >
                      <span
                        className="codicon codicon-chevron-down"
                        onMouseEnter={[Function]}
                        onMouseLeave={[Function]}
                      />
                    </li>
                  </ul>
                </div>
              </header>
              <div
                className="mo-notification__body"
              />
            </div>
          </a>
        </div>
        <div
          className="mo-statusBar__item"
          id="MoEditorInfo"
        >
          <a
            onClick={[Function]}
            tabIndex={-1}
            title="Go to Line/Column"
          >
            <span>
              Ln 0, Col 0
            </span>
          </a>
        </div>
      </div>
    </div>
  </div>
</DocumentFragment>
`;
